openquake/hazardlib/source/kite_fault.py

```diff
@@ -0,0 +1,239 @@
+# The Hazard Library
+# Copyright (C) 2012-2020 GEM Foundation
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU Affero General Public License as
+# published by the Free Software Foundation, either version 3 of the
+# License, or (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU Affero General Public License for more details.
+#
+# You should have received a copy of the GNU Affero General Public License
+# along with this program. If not, see <http://www.gnu.org/licenses/>.
+"""
+Module :mod:`openquake.hazardlib.source.kite_fault` defines
+:class:`KiteFaultSource`.
+"""
+
+import numpy as np
+from typing import Tuple, Optional
+from openquake.hazardlib.geo.mesh import Mesh
+from openquake.hazardlib.source.base import ParametricSeismicSource
+from openquake.hazardlib.geo.surface.kite_fault import KiteSurface
+from openquake.hazardlib.source.rupture import ParametricProbabilisticRupture \
+    as ppr
+
+
+class KiteFaultSource(ParametricSeismicSource):
+    """
+    Kite fault source
+    """
+
+    MODIFICATIONS = {}
+
+    def __init__(self, source_id, name, tectonic_region_type, mfd,
+                 rupture_mesh_spacing, magnitude_scaling_relationship,
+                 rupture_aspect_ratio, temporal_occurrence_model,
+                 # kite fault specific parameters
+                 profiles, floating_x_step,
+                 floating_y_step, rake, profiles_sampling=None):
+        super().__init__(
+            source_id, name, tectonic_region_type, mfd, rupture_mesh_spacing,
+            magnitude_scaling_relationship, rupture_aspect_ratio,
+            temporal_occurrence_model)
+
+        # TODO add checks
+        self.profiles = profiles
+        if profiles_sampling is None:
+            self.profiles_sampling = (rupture_mesh_spacing /
+                                      rupture_aspect_ratio)
+        self.floating_x_step = floating_x_step
+        self.floating_y_step = floating_y_step
+        self.rake = rake
+
+        min_mag, max_mag = self.mfd.get_min_max_mag()
+
+    @property
+    def surface(self) -> KiteSurface:
+        """
+        :returns:
+            The surface of the fault
+        """
+        # Get the surface of the fault
+        # TODO we must automate the definition of the idl parameter
+        return KiteSurface.from_profiles(self.profiles,
+                                              self.profiles_sampling,
+                                              self.rupture_mesh_spacing,
+                                              idl=False, align=False)
+
+    def count_ruptures(self) -> int:
+        """
+        :returns:
+            The number of ruptures that this source generates
+        """
+        return len([r for r in self.iter_ruptures()])
+
```

Comment:
This is cheating. The point of `count_ruptures` is to have an algorithm to count the ruptures without having to generate all of them. As it is now it will work but the preprocessing can be slow.

Reply:
I know. But the rupture generation process is fast. Given the mechanism used to generate ruptures, unfortunately there is no other (quicker) mechanism.

Reply:
At least you can avoid keeping all ruptures in memory with
```return sum(1 for r in  self.iter_ruptures())```